core/src/eval/mod.rs

```diff
@@ -943,11 +943,14 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
     ///   fields and array elements, we keep evaluating subsequent elements even if one
     ///   fails.
     /// - We only return the accumulated errors; we don't return the eval'ed term.
-    pub fn eval_permissive(&mut self, rt: RichTerm) -> Vec<EvalError> {
+    /// - We support a recursion limit, to limit the number of times we recurse into
+    ///   arrays or records.
+    pub fn eval_permissive(&mut self, rt: RichTerm, recursion_limit: u64) -> Vec<EvalError> {
```

Comment:
Nitpick: should we use `usize` instead of `u64`? I guess in practice `u16` is already amply sufficient, but you might prefer to use a word-aligned value. Using `u64` sounds like we need the guarantee to have larger values than `u32::MAX`.